src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/Semantics/ExprFactory.cs

```diff
@@ -185,142 +146,93 @@ public ExprUserDefinedConversion CreateUserDefinedConversion(Expr arg, Expr call
-
-            CType type = null;
-            if (pLocal != null)
-            {
-                type = pLocal.GetType();
-            }
-
             ExprLocal rval = new ExprLocal();
-            rval.Kind = ExpressionKind.EK_LOCAL;
-            rval.Type = type;
             rval.Flags = nFlags;
             rval.Local = pLocal;
-            Debug.Assert(rval != null);
-            return (rval);
+            return rval;
         }
 
         public ExprThisPointer CreateThis(LocalVariableSymbol pLocal, bool fImplicit)
         {
             Debug.Assert(pLocal == null || pLocal.isThis);
 
-            CType type = null;
-            if (pLocal != null)
-            {
-                type = pLocal.GetType();
-            }
-
             EXPRFLAG flags = EXPRFLAG.EXF_CANTBENULL;
             if (fImplicit)
             {
                 flags |= EXPRFLAG.EXF_IMPLICITTHIS;
             }
-            if (type != null && type.isStructType())
+            if (pLocal != null && pLocal.GetType().isStructType())
             {
                 flags |= EXPRFLAG.EXF_LVALUE;
             }
 
             ExprThisPointer rval = new ExprThisPointer();
-            rval.Kind = ExpressionKind.EK_THISPOINTER;
-            rval.Type = type;
             rval.Flags = flags;
             rval.Local = pLocal;
-            Debug.Assert(rval != null);
             return (rval);
         }
 
         public ExprBoundLambda CreateAnonymousMethod(AggregateType delegateType)
         {
             Debug.Assert(delegateType == null || delegateType.isDelegateType());
-            ExprBoundLambda rval = new ExprBoundLambda();
-            rval.Kind = ExpressionKind.EK_BOUNDLAMBDA;
-            rval.Type = delegateType;
-            rval.Flags = 0;
-            Debug.Assert(rval != null);
-            return (rval);
+            ExprBoundLambda rval = new ExprBoundLambda(delegateType);
+            return rval;
         }
 
         public ExprUnboundLambda CreateLambda()
         {
-            CType type = GetTypes().GetAnonMethType();
-
-
-            ExprUnboundLambda rval = new ExprUnboundLambda();
-            rval.Kind = ExpressionKind.EK_UNBOUNDLAMBDA;
-            rval.Type = type;
-            rval.Flags = 0;
-            Debug.Assert(rval != null);
-            return (rval);
+            ExprUnboundLambda rval = new ExprUnboundLambda(GetTypes().GetAnonMethType());
+            return rval;
         }
 
         public ExprHoistedLocalExpr CreateHoistedLocalInExpression(ExprLocal localToHoist)
         {
             Debug.Assert(localToHoist != null);
-            ExprHoistedLocalExpr rval = new ExprHoistedLocalExpr();
-            rval.Kind = ExpressionKind.EK_HOISTEDLOCALEXPR;
-            rval.Type = GetTypes().GetOptPredefAgg(PredefinedType.PT_EXPRESSION).getThisType();
-            rval.Flags = 0;
-            return rval;
+            return new ExprHoistedLocalExpr(GetTypes().GetOptPredefAgg(PredefinedType.PT_EXPRESSION).getThisType());
```

Comment:
Possibly these cases that go through `GetTypes()` can have that logic moved into something static, as it **seems** the types factory doesn't pass any dependencies into these predefined types and so they could become static and that would pass down to simplifying these `Expr` types, but I'm not examining that yet.